File: outbox/cdc/orderservice/src/main/java/com/varunu28/orderservice/service/OrderService.java
package com.varunu28.orderservice.service;

import com.varunu28.orderservice.dto.CreateOrderDto;
import com.varunu28.orderservice.dto.OrderDto;
import com.varunu28.orderservice.model.Order;
import com.varunu28.orderservice.model.Outbox;
import com.varunu28.orderservice.model.OutboxEventType;
import com.varunu28.orderservice.repository.OrderRepository;
import com.varunu28.orderservice.repository.OutboxRepository;
import jakarta.transaction.Transactional;
import org.springframework.stereotype.Service;

@Service
public class OrderService {

    private final OrderRepository orderRepository;
    private final OutboxRepository outboxRepository;

    public OrderService(OrderRepository orderRepository, OutboxRepository outboxRepository) {
        this.orderRepository = orderRepository;
        this.outboxRepository = outboxRepository;
    }

    @Transactional
    public OrderDto createOrder(CreateOrderDto request) {
        Order order = new Order(request.customerName(), request.amount());
        Order savedOrder = orderRepository.save(order);
        var payload = """
            {
                "order_id": %d,
                "customer_name": "%s",
                "amount": %s
            }
            """.formatted(savedOrder.getId(), savedOrder.getCustomerName(), savedOrder.getAmount());
        Outbox outbox = new Outbox(OutboxEventType.ORDER_CREATED.getValue(), payload);
        outboxRepository.save(outbox);
        return OrderDto.from(savedOrder);
    }
}

File: saga/order-service/src/main/kotlin/com/saga/orderservice/OrderService.kt
package com.saga.orderservice

import com.saga.orderservice.kafka.KafkaTopics.ORDER_CREATED_TOPIC
import com.saga.orderservice.kafka.OrderCreatedEvent
import com.saga.orderservice.saga.OrderWorkflowService
import jakarta.transaction.Transactional
import kotlinx.serialization.json.Json
import org.springframework.data.repository.findByIdOrNull
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.stereotype.Service
import java.math.BigInteger
import java.time.LocalDateTime
import java.util.*

@Service
class OrderService(
    private val orderRepository: OrderRepository,
    private val orderWorkflowService: OrderWorkflowService,
    private val kafkaTemplate: KafkaTemplate<String, String>
) {

    @Transactional
    fun createOrder(request: CreateOrderRequest): BigInteger {
        val now = LocalDateTime.now()
        val order = Order(
            customerName = request.customerName,
            amount = request.amount,
            inventoryCount = request.inventoryCount,
            status = "CREATED",
            createdAt = now,
            updatedAt = now
        )
        val savedOrder = orderRepository.save(order)
        val orderId = savedOrder.id ?: BigInteger.ZERO
        val idempotencyKey = UUID.randomUUID().toString()
        val orderCreatedEvent = OrderCreatedEvent(
            orderId.toString(), savedOrder.inventoryCount, savedOrder.amount, savedOrder.customerName, idempotencyKey
        )
        val eventString = Json.encodeToString(orderCreatedEvent)
        kafkaTemplate.send(ORDER_CREATED_TOPIC, eventString)
        return orderId
    }

    fun getOrderById(id: BigInteger): Order {
        return orderRepository.findByIdOrNull(id) ?: throw OrderNotFoundException()
    }

    fun processOrderCreatedEvent(
        orderId: BigInteger,
        inventoryCount: Int,
        amount: Double,
        customerName: String,
        idempotencyKey: String
    ) {
        orderWorkflowService.processOrder(
            orderId = orderId,
            inventoryCount = inventoryCount,
            customerName = customerName,
            amount = amount,
            idempotencyKey = idempotencyKey
        )
    }

    @Transactional
    fun processOrderFailedEvent(orderId: BigInteger) {
        val order = orderRepository.findByIdOrNull(orderId) ?: return
        order.status = "FAILED"
        order.updatedAt = LocalDateTime.now()
        orderRepository.save(order)
    }

    @Transactional
    fun processOrderSucceededEvent(orderId: BigInteger) {
        val order = orderRepository.findByIdOrNull(orderId) ?: return
        order.status = "SUCCEEDED"
        order.updatedAt = LocalDateTime.now()
        orderRepository.save(order)
    }
}